lib/Analysis/DependenceAnalysis.cpp

```diff
@@ -0,0 +1,72 @@
+//===- DependenceAnalysis.cpp - memory dependence analyses ----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements methods that perform analysis involving memory access
+// dependences.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Analysis/DependenceAnalysis.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Transforms/LoopUtils.h"
+#include <mlir/Analysis/AffineStructures.h>
+#include <mlir/Dialect/Affine/IR/AffineMemoryOpInterfaces.h>
+
+using namespace mlir;
+using namespace circt::analysis;
+
+/// Helper to iterate through memory operation pairs and check for dependences
+/// at a given loop nesting depth.
+static void checkMemrefDependence(SmallVectorImpl<Operation *> &memoryOps,
+                                  unsigned depth,
+                                  MemoryDependenceResult &results) {
+  for (auto *source : memoryOps) {
+    results[source] = SmallVector<MemoryDependence>();
+    for (auto *destination : memoryOps) {
+      if (source == destination)
+        continue;
+
+      MemRefAccess src(source);
+      MemRefAccess dst(destination);
+      FlatAffineValueConstraints dependenceConstraints;
+      SmallVector<DependenceComponent, 2> depComps;
+      DependenceResult result = checkMemrefAccessDependence(
+          src, dst, depth, &dependenceConstraints, &depComps, true);
+
+      results[source].emplace_back(destination, result.value, depComps);
+    }
+  }
+}
+
+/// MemoryDependenceAnalysis traverses any AffineForOps in the FuncOp body and
+/// checks for memory access dependences. Results are captured in a
+/// MemoryDependenceResult, which can by queried by Operation.
+circt::analysis::MemoryDependenceAnalysis::MemoryDependenceAnalysis(
+    mlir::FuncOp funcOp) {
+  // Collect affine loops grouped by nesting depth.
+  std::vector<SmallVector<AffineForOp, 2>> depthToLoops;
+  mlir::gatherLoops(funcOp, depthToLoops);
+
+  // Collect load and store operations to check.
+  SmallVector<Operation *> memoryOps;
+  funcOp.walk([&](Operation *op) {
+    if (isa<AffineReadOpInterface, AffineWriteOpInterface>(op))
+      memoryOps.push_back(op);
+  });
+
+  // For each depth, check memref accesses.
+  for (unsigned depth = 1, e = depthToLoops.size(); depth <= e; ++depth)
+    checkMemrefDependence(memoryOps, depth, results);
+}
+
+/// Returns the dependences, if any, that originate from the given Operation.
```

Comment:
I'm doing it the other way round in the [scheduling::Problem](https://github.com/llvm/circt/blob/4379fab383edf6da8bb86d04ae6a41970e620b9f/include/circt/Scheduling/Problems.h#L166-L176) class. Currently, your query gives you `op`'s outgoing dependences, which reference other ops that need to happen after `op`. My rationale for attaching incoming dependences to the respective destination ops would be that then this query would read "give me the ops that need to happen before `op` can execute", similar to value operands, which I find more intuitive. Functionally, it doesn't make a difference, though. WDYT?